### PR TITLE
execute saved query

### PR DESCRIPTION
**Pull request by bheni**
Implements:
```
dolt sql -x <saved_query_name>
dolt sql --list-saved
```
Changes
 * `dolt sql -s <name> -q <query>` now saves the query with id = name

### CODE REVIEW

File: go/cmd/dolt/commands/sql.go [hunk 1]
```diff
@@ -314,11 +359,37 @@ func validateSqlArgs(apr *argparser.ArgParseResults) error {
...
+			return errhand.BuildDError("Invalid Argument: --list-saved is not compatible with --executed|x").Build()
+		} else if query {
+			return errhand.BuildDError("Invalid Argument: --list-saved is not compatible with --query|-q").Build()
+		} else if msg {
+			return errhand.BuildDError("Invalid Argument: --list-saved is not compatible with --message|-m").Build()
+		} else if save {
+			return errhand.BuildDError("Invalid Argument: --list-saved is not compatible with --save|-s").Build()
+		}
+	}
+
```
Comment (bheni) [inline go/cmd/dolt/commands/sql.go]: Gotta be a better way to do this.

File: go/cmd/dolt/commands/sql.go [hunk 2]
```diff
@@ -63,7 +63,10 @@ var sqlDocs = cli.CommandDocumentationContent{
 		"results, then exits.\n" +
 		"\n" +
 		"By default, {{.EmphasisLeft}}-q{{.EmphasisRight}} executes a single statement. To execute multiple SQL " +
-		"statements separated by semicolons, use -b to enable batch mode.\n" +
+		"statements separated by semicolons, use {{.EmphasisLeft}}-b{{.EmphasisRight}} to enable batch mode. Queries can" +
+		"be saved with {{.EmphasisLeft}}-s{{.EmphasisRight}}.\n" +
+		"\n" +
+		"Alternatively {{.EmpahasisLeft}}-x{{.EmphasisRight}} can be useod to executed a saved query by name\n" +
```
Comment (zachmu) [inline go/cmd/dolt/commands/sql.go]: used to execute

Comment (zachmu) [reply go/cmd/dolt/commands/sql.go]: and end sentence with .

File: go/cmd/dolt/commands/sql.go [hunk 3]
```diff
@@ -82,16 +85,20 @@ var sqlDocs = cli.CommandDocumentationContent{
 		"-q {{.LessThan}}query;query{{.GreaterThan}} -b",
 		"-q {{.LessThan}}query{{.GreaterThan}} -r {{.LessThan}}result format{{.GreaterThan}}",
 		"-q {{.LessThan}}query{{.GreaterThan}} -s {{.LessThan}}name{{.GreaterThan}} -m {{.LessThan}}message{{.GreaterThan}}",
+		"-x {{.LessThan}}name{{.GreaterThan}}",
+		"-list-saved",
```
Comment (zachmu) [inline go/cmd/dolt/commands/sql.go]: --list-saved

File: go/cmd/dolt/commands/sql.go [hunk 4]
```diff
@@ -119,6 +126,8 @@ func (cmd SqlCmd) createArgParser() *argparser.ArgParser {
 	ap.SupportsString(queryFlag, "q", "SQL query to run", "Runs a single query and exits")
 	ap.SupportsString(formatFlag, "r", "result output format", "How to format result output. Valid values are tabular, csv. Defaults to tabular. ")
 	ap.SupportsString(saveFlag, "s", "saved query name", "Used with --query, save the query to the query catalog with the name provided. Saved queries can be examined in the dolt_query_catalog system table.")
+	ap.SupportsString(executeFlag, "x", "saved query name", "Executes a saved query with the given name")
+	ap.SupportsFlag(listSavedFlag, "", "Lists all saved queries")
```
Comment (zachmu) [inline go/cmd/dolt/commands/sql.go]: no -l?

File: go/cmd/dolt/commands/sql.go (hunk 1 above)
Comment (zachmu) [reply go/cmd/dolt/commands/sql.go]: Maybe, but this works fine

File: go/cmd/dolt/commands/sql.go [hunk 5]
```diff
@@ -155,15 +169,46 @@ func (cmd SqlCmd) Exec(ctx context.Context, commandStr string, args []string, dE
 	saveMessage := apr.GetValueOrDefault(messageFlag, "")
 	saveName := apr.GetValueOrDefault(saveFlag, "")
 
-	err := validateSqlArgs(apr)
-	if err != nil {
-		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+	_, forceBatchMode := apr.GetValue(batchFlag)
+
+	query, queryOK := apr.GetValue(queryFlag)
+	if !queryOK {
```
Comment (zachmu) [inline go/cmd/dolt/commands/sql.go]: You already validated this in the validate method, no need for this check here. Just use exOk directly

File: go/cmd/dolt/commands/sql.go [hunk 6]
```diff
@@ -155,15 +169,46 @@ func (cmd SqlCmd) Exec(ctx context.Context, commandStr string, args []string, dE
...
+
+		if exOk {
+			sq, err := dsqle.RetrieveFromQueryCatalog(ctx, root, savedQueryName)
+
+			if err != nil {
+				return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+			}
+
+			query = sq.Query
+			queryOK = true
```
Comment (zachmu) [inline go/cmd/dolt/commands/sql.go]: Don't like this kind of fallthrough logic, would much prefer extracting the logic used in the --query path into a method and invoking it here with an exit

File: go/cmd/dolt/commands/sql.go [hunk 7]
```diff
@@ -155,15 +169,46 @@ func (cmd SqlCmd) Exec(ctx context.Context, commandStr string, args []string, dE
...
+			if err != nil {
+				verr := errhand.BuildDError("error: Failed to read from repository.").AddCause(err).Build()
+				return HandleVErrAndExitCode(verr, usage)
+			}
+
+			if !hasQC {
+				return 0
+			}
+
+			query = "SELECT * FROM " + doltdb.DoltQueryCatalogTableName
```
Comment (zachmu) [inline go/cmd/dolt/commands/sql.go]: Same comment about fallthrough

File: go/libraries/doltcore/sqle/query_catalog_table.go [hunk 8]
```diff
@@ -75,62 +118,127 @@ func createQueryCatalogIfNotExists(ctx context.Context, root *doltdb.RootValue)
...
-	root, err := createQueryCatalogIfNotExists(ctx, root)
+func NewQueryCatalogEntryWithRandID(ctx context.Context, root *doltdb.RootValue, name, query, description string) (SavedQuery, *doltdb.RootValue, error) {
+	uid, err := uuid.NewRandom()
 	if err != nil {
-		return nil, err
+		return SavedQuery{}, nil, err
 	}
 
-	tbl, _, err := root.GetTable(ctx, doltdb.DoltQueryCatalogTableName)
+	// Use the last 24 hex digits of the uuid for the ID.
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/query_catalog_table.go]: Last 12

Comment (bheni) [reply go/libraries/doltcore/sqle/query_catalog_table.go]: fixed.

File: go/cmd/dolt/commands/sql.go (hunk 2 above)
Comment (bheni) [reply go/cmd/dolt/commands/sql.go]: fixed

File: go/cmd/dolt/commands/sql.go (hunk 3 above)
Comment (bheni) [reply go/cmd/dolt/commands/sql.go]: fixed

File: go/cmd/dolt/commands/sql.go (hunk 4 above)
Comment (bheni) [reply go/cmd/dolt/commands/sql.go]: sure

File: go/cmd/dolt/commands/sql.go (hunk 5 above)
Comment (bheni) [reply go/cmd/dolt/commands/sql.go]: this got changed when I addressed your other comments.

File: go/cmd/dolt/commands/sql.go (hunk 6 above)
Comment (bheni) [reply go/cmd/dolt/commands/sql.go]: done

File: go/cmd/dolt/commands/sql.go (hunk 7 above)
Comment (bheni) [reply go/cmd/dolt/commands/sql.go]: done

File: go/cmd/dolt/commands/sql.go [hunk 9]
```diff
@@ -164,52 +164,52 @@ func (cmd SqlCmd) Exec(ctx context.Context, commandStr string, args []string, dE
...
 		}
+
+		query := "SELECT * FROM " + doltdb.DoltQueryCatalogTableName
+		return execQuery(ctx, dEnv, root, query, "", "", forceBatchMode, format, usage)
 	}
 
+	return execQuery(ctx, dEnv, root, "", "", "", forceBatchMode, format, usage)
+}
+
+func execQuery(ctx context.Context, dEnv *env.DoltEnv, root *doltdb.RootValue, query, saveName, saveMessage string, forceBatchMode bool, format resultFormat, usage cli.UsagePrinter) int {
```
Comment (zachmu) [inline go/cmd/dolt/commands/sql.go]: This should be three separate functions invoked from the top level Exec function, one for each major code path.

execQuery should run a single query and exit
execBatchModeQueries should run batch mode
execShell should run the shell

**Pull request merged**